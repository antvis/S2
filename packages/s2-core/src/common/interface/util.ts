export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
