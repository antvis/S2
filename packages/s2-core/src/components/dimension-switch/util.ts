import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
