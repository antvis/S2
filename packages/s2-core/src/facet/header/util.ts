import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
