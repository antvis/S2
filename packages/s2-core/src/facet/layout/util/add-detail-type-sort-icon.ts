import { Group } from '@antv/g-canvas';
import * as _ from 'lodash';
import { GuiIcon } from '../../../common/icons';
import { DEFAULT_PADDING, KEY_LIST_SORT } from '../../../common/constant';
import { SpreadSheet } from '../../..';

// 排序按钮的宽度
export const SORT_ICON_SIZE = 8;

export const getIconType = (key: string, spreadsheet: SpreadSheet) => {
  /*
  currentSortKey 存储的点击的某个field对应的升序还是降序
  currentSortKey = {
   [field]: up/down
  }
   */
  const currentSortKey = spreadsheet.store.get('currentSortKey', {});
  let upIconType = 'SortUp';
  let downIconType = 'SortDown';
  if (currentSortKey && _.has(currentSortKey, key)) {
    // 有配置,当前点击的过的key(某个维度)
    if (_.get(currentSortKey, key) === 'up') {
      // 点击过此维度的up
      upIconType = 'SortUpSelected';
    } else {
      // 点击过此维度的down
      downIconType = 'SortDownSelected';
    }
  }

  return {
    upIconType,
    downIconType,
  };
};

export const renderIcon = (
  parent: Group,
  spreadsheet: SpreadSheet,
  x: number,
  y: number,
  type: string,
  key: string,
  sortKeyVal: string,
  sortType: string,
) => {
  const icon = new GuiIcon({
    type,
    x,
    y,
    width: SORT_ICON_SIZE,
    height: SORT_ICON_SIZE,
  });

  icon.on('click', () => {
    const selected = icon.get('selected');
    let currentSortKey = {};
    if (!selected) {
      currentSortKey = {
        [key]: sortKeyVal,
      };
    }
    spreadsheet.store.set('currentSortKey', currentSortKey);
    icon.set('selected', !selected);
    spreadsheet.emit(KEY_LIST_SORT, {
      sortFieldId: key,
      sortMethod: selected ? '' : sortType,
    });
    spreadsheet.render(false);
  });

  parent.add(icon);
};

export const renderDetailTypeSortIcon = (
  parent: Group,
  spreadsheet: SpreadSheet,
  textX: number,
  textY,
  key,
) => {
  const iconType = getIconType(key, spreadsheet);

  renderIcon(
    parent,
    spreadsheet,
    textX,
    textY - DEFAULT_PADDING * 2,
    iconType.upIconType,
    key,
    'up',
    'ASC',
  );
  renderIcon(
    parent,
    spreadsheet,
    textX,
    textY - DEFAULT_PADDING / 2,
    iconType.downIconType,
    key,
    'down',
    'DESC',
  );
};
