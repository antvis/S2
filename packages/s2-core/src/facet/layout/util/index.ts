export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
