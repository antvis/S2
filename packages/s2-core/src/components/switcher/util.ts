import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
