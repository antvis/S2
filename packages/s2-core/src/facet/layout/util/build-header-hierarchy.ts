import {
  size,
  includes,
  isEmpty,
  each,
  merge,
  isBoolean,
  remove,
} from 'lodash';
import { Pivot } from '../../../data-set';
import { i18n } from '../../../common/i18n';
import { SpreadsheetFacetCfg } from '../../../common/interface';
import { Hierarchy } from '../hierarchy';
import { Node } from '../node';
import TotalClass from '../total-class';
import { canNodeBeExpanded } from './can-node-be-expanded';
import getDimsCondition from './get-dims-condition-by-node';
import { EXTRA_FIELD } from '../../../common/constant';
import { handleHideNodes } from './handle-hide-nodes';
import { generateId } from './generate-id';

function addTotalsNodes(
  pivot: Pivot,
  field: string,
  fieldValues: (string | TotalClass)[],
  isSubTotals: boolean,
) {
  const totalsConfig = pivot.getTotalsConfig(field);
  const option = isSubTotals
    ? totalsConfig.showSubTotals
    : totalsConfig.showGrandTotals;
  if (option) {
    // eslint-disable-next-line no-nested-ternary
    const func = isSubTotals
      ? totalsConfig.reverseSubLayout
        ? 'unshift'
        : 'push'
      : totalsConfig.reverseLayout
      ? 'unshift'
      : 'push';
    const value = isSubTotals
      ? new TotalClass(totalsConfig.subLabel, true)
      : new TotalClass(totalsConfig.label, false, true);
    fieldValues[func](value);
  }
}

function handleTotals(
  field: string,
  fields: string[],
  index,
  fieldValues: (string | TotalClass)[],
  pivot: Pivot,
) {
  // put all totals nodes in appropriate place
  if (field !== fields[0]) {
    // sub totals from 1 to length - 1
    const previousField = fields[index - 1];
    // NOTE: if current field's only has one child, we don't need sub totals
    if (size(fieldValues) > 1) {
      // [a, b, c] (start from second index, b -> a, c -> b)
      addTotalsNodes(pivot, previousField, fieldValues, true);
    }
  } else if (field === fields[0]) {
    // grand totals in first field level
    // check if there are grand totals in first field
    addTotalsNodes(pivot, fields[0], fieldValues, false);
  }
}

export default function buildHeaderHierarchy(
  pivot: Pivot,
  parent: Node,
  field: string,
  fields: string[],
  cfg: SpreadsheetFacetCfg,
  hierarchy: Hierarchy,
) {
  const index = fields.indexOf(field);
  const key = field;
  const query = getDimsCondition(parent, true);
  const fieldValues: (string | TotalClass)[] = pivot.getDimValues(key, query);

  // when cols/values not exist,we need show table structure with empty body
  if (key === EXTRA_FIELD && includes(cfg.cols, key) && isEmpty(fieldValues)) {
    // add empty measure node
    fieldValues.push('');
  }

  // handle handle nodes by node query
  handleHideNodes(fieldValues, field, cfg);

  // handle totals nodes
  handleTotals(field, fields, index, fieldValues, pivot);
  const { collapsedCols } = cfg;
  // generate all nodes(normal and totals) of this field
  each(fieldValues, (col) => {
    const isTotal = col instanceof TotalClass;
    let value;
    if (isTotal) {
      value = i18n((col as TotalClass).label);
    } else {
      value = col;
    }
    const nodeQuery = merge({}, query, { [field]: value });
    const id = generateId(parent.id, value, cfg);
    if (!id) return;
    const isCollapsed = isBoolean(collapsedCols[id])
      ? collapsedCols[id]
      : false;
    const node = new Node({
      id,
      key,
      value,
      level: index,
      field,
      parent,
      isTotals: isTotal,
      isCollapsed,
      isSubTotals: isTotal ? (col as TotalClass).isSubTotals : false,
      isGrandTotals: isTotal ? (col as TotalClass).isGrandTotals : false,
      hierarchy,
      query: nodeQuery,
      spreadsheet: cfg.spreadsheet,
    });

    if (cfg.hierarchy) {
      const another = cfg.hierarchy(cfg.spreadsheet, node);
      const push = another?.push;
      if (another.nodes.length > 0) {
        // 存在节点，按push 或者 unshift插入
        if (push) {
          hierarchy.pushNode(node);
        }
        each(another.nodes, (v) => {
          hierarchy.pushNode(v);
        });
        if (!push) {
          // new node insert before current
          hierarchy.pushNode(node);
          // adjust the index of current node
          remove(parent.children, (v) => v === node);
          parent.children.push(node);
        }
      } else {
        // 不存在节点，正常插入
        hierarchy.pushNode(node);
      }
    } else {
      // no extra node exist
      hierarchy.pushNode(node);
    }

    if (index > hierarchy.maxLevel) {
      hierarchy.sampleNodesForAllLevels.push(node);
      hierarchy.sampleNodeForLastLevel = node;
      hierarchy.maxLevel = index;
    }

    if (index === fields.length - 1 || isTotal || isCollapsed) {
      node.isLeaf = true;
      hierarchy.pushIndexNode(node);
      node.rowIndex = hierarchy.getIndexNodes().length - 1;
    }
    if (canNodeBeExpanded(index < fields.length - 1, isCollapsed, isTotal)) {
      buildHeaderHierarchy(
        pivot,
        node,
        fields[index + 1],
        fields,
        cfg,
        hierarchy,
      );
    }
  });
}
