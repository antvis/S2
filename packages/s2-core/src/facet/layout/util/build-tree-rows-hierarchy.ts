import { each, mix } from '@antv/util';
import * as _ from 'lodash';
import { Pivot, SpreadDataSet } from '../../../data-set';
import { i18n } from '../../../common/i18n';
import { SpreadsheetFacetCfg } from '../../../common/interface';
import { Hierarchy } from '../hierarchy';
import { Node } from '../node';
import TotalClass from '../total-class';
import { canNodeBeExpanded } from './can-node-be-expanded';
import getDimsCondition from './get-dims-condition-by-node';
import { handleHideNodes } from './handle-hide-nodes';
import { handleKeepOnlyNodes } from './handle-keep-only-nodes';
import { reArrangeFieldValues } from './re-arrange-field-values';
import findNodeExtraCfg from '../../../utils/find-node-extra-cfg';

export interface TreeParams {
  pivot: Pivot;
  parent: Node;
  field: string;
  fields: string[];
  cfg: SpreadsheetFacetCfg;
  hierarchy: Hierarchy;
  dataSet: SpreadDataSet;
  // check if this node in collapse node
  inCollapseNode: boolean;
}

export default function buildTreeRowsHierarchy(treeParams: TreeParams) {
  const {
    pivot,
    parent,
    field,
    fields,
    cfg,
    hierarchy,
    dataSet,
    inCollapseNode,
  } = treeParams;
  const index = fields.indexOf(field);
  const key = field;
  const query = getDimsCondition(parent, true);
  let fieldValues: (string | TotalClass)[] = pivot.getDimValues(key, query);
  // re-arrange field values by subGrand values
  reArrangeFieldValues(query, field, fieldValues, dataSet);
  // custom re-arrange
  if (cfg.layoutArrange) {
    fieldValues = cfg.layoutArrange(
      cfg.spreadsheet,
      parent,
      field,
      fieldValues as string[],
    );
  }
  const totalsConfig = pivot.getTotalsConfig(field);

  // handle keep only node in row/col by ids
  handleKeepOnlyNodes(pivot, fieldValues, field, parent.id);
  // handle handle nodes by node query
  handleHideNodes(pivot, fieldValues, field, parent.id);

  // tree mode only has grand totals, but if there are subTotals configs, it will
  // display in cross-area cell
  if (field === fields[0] && totalsConfig.showGrandTotals) {
    const func = totalsConfig.reverseLayout ? 'unshift' : 'push';
    fieldValues[func](new TotalClass(totalsConfig.label, false, true));
  }

  const { collapsedRows, hierarchyCollapse, values } = cfg;

  each(fieldValues, (col) => {
    const isTotal = col instanceof TotalClass;
    let value;
    let label;
    if (isTotal) {
      value = i18n((col as TotalClass).label);
      label = value;
    } else if (cfg.spreadsheet.isStrategyMode()) {
      value = col;
      // 在此必须区分label和value，自动解析的场景下 value=label，但是在决策模式下如果
      // 用value去替代label 会导致query条件出错
      // 决策模式下的度量值文本名字设置（度量格式化的名字+衍生指标）
      label = dataSet.getFieldName(col);
    } else {
      value = col;
      label = value;
    }
    const id = `${parent.id}-${value}`;

    // check if current node is collapsed first, or use hierarchy config
    let isCollapse = _.isBoolean(collapsedRows[id])
      ? collapsedRows[id]
      : hierarchyCollapse;
    if (isTotal) {
      // 总计用户不会有收缩状态
      isCollapse = false;
    }
    // 处理决策模式下，初始化节点的收缩状态，一次性！！
    if (_.isBoolean(collapsedRows[id]) || hierarchyCollapse) {
      // 有操作后节点的情况下，需要以操作的为准，isCollapse不变
    } else {
      // 没有操作的过节点的情况下，默认以配置为准
      const extra = findNodeExtraCfg(values, { [field]: value });
      // 必须不为空
      if (extra && !_.isEmpty(label)) {
        isCollapse = extra.collapse;
      }
    }
    const node = new Node({
      id,
      key,
      label,
      value,
      level: index,
      parent,
      field,
      isTotals: isTotal,
      isCollapsed: isCollapse,
      hierarchy,
      query: mix({}, query, { [field]: value }),
      inCollapseNode,
      spreadsheet: cfg.spreadsheet,
    });
    if (cfg.hierarchy) {
      const another = cfg.hierarchy(cfg.spreadsheet, node);
      const push = another?.push;
      if (another.nodes.length > 0) {
        // 存在节点，按push 或者 unshift插入
        if (push) {
          hierarchy.pushNode(node);
        }
        _.each(another.nodes, (v) => {
          hierarchy.pushNode(v);
        });
        if (!push) {
          // new node insert before current
          hierarchy.pushNode(node);
          // adjust the index of current node
          _.remove(parent.children, (v) => value === node);
          parent.children.push(node);
        }
      } else {
        // 不存在节点，正常插入
        hierarchy.pushNode(node);
      }
    } else {
      // no extra node exist
      hierarchy.pushNode(node);
    }

    if (index > hierarchy.maxLevel) {
      hierarchy.sampleNodesForAllLevels.push(node);
      hierarchy.sampleNodeForLastLevel = node;
      hierarchy.maxLevel = index;
    }

    if (index === fields.length - 1 || isTotal) {
      node.isLeaf = true;
      hierarchy.pushIndexNode(node);
      node.rowIndex = hierarchy.getIndexNodes().length - 1;
    }
    // collapsed node and totals don't have branches
    /**
     * Tree rows hierarchy now has different ways to handle node
     * Old strategy: use isCollapse to determine if render next level nodes(if true, we dont have
     * nodes in next level)
     * New strategy: use inCollapseNode, the next level's nodes still exist, but we "collapse" them
     * by height === 0, check {@hideRowColumnsByFields#l30}
     */
    if (canNodeBeExpanded(index < fields.length - 1, false, isTotal)) {
      buildTreeRowsHierarchy({
        pivot,
        parent: node,
        field: fields[index + 1],
        fields,
        cfg,
        hierarchy,
        dataSet,
        inCollapseNode: isCollapse || inCollapseNode,
      } as TreeParams);
    }
  });
}
