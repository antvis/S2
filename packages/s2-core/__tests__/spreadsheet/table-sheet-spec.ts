import { getContainer, getMockData, sleep } from 'tests/util/helpers';
import { TableSheet, S2Options, S2DataConfig, ResizeType } from '@/index';

const data = getMockData(
  '../../../s2-react/__tests__/data/tableau-supermarket.csv',
);

const columns = [
  'order_id',
  'order_date',
  'ship_date',
  'express_type',
  'customer_name',
  'customer_type',
  'city',
  'province',
  'counter',
  'area',
  'type',
  'sub_type',
  'product_name',
  'sale_amt',
  'count',
  'discount',
  'profit',
];

const meta = [
  {
    field: 'count',
    name: '销售个数',
  },
  {
    field: 'profit',
    name: '利润',
    formatter: (v: number) => `${v}元`,
  },
];

const newLineText = `1\t\n2`;

const dataCfg: S2DataConfig = {
  fields: {
    columns,
  },
  meta,
  data: data.map((e) => ({ ...e, express_type: newLineText })),
  sortParams: [
    {
      sortFieldId: 'count',
      sortMethod: 'DESC',
    },
    {
      sortFieldId: 'profit',
      sortMethod: 'ASC',
    },
  ],
};

const options: S2Options = {
  width: 800,
  height: 600,
  showSeriesNumber: true,
  placeholder: '',
  style: {
    layoutWidthType: 'compact',
    cellCfg: {
      height: 32,
    },
    device: 'pc',
  },
  interaction: {
    enableCopy: true,
    hoverHighlight: false,
    linkFields: ['order_id', 'customer_name'],
    hiddenColumnFields: ['order_date'],
    resize: {
      rowResizeType: ResizeType.CURRENT,
    },
  },
  frozenRowCount: 2,
  frozenColCount: 2,
  frozenTrailingColCount: 2,
  frozenTrailingRowCount: 2,
  showDefaultHeaderActionIcon: true,
  tooltip: {
    operation: {
      hiddenColumns: true,
    },
  },
};

describe('TableSheet normal spec', () => {
  test('scrollWithAnimation with duration and callback', async () => {
    const s2 = new TableSheet(getContainer(), dataCfg, options);
    s2.render();

    const onScrollFinish = jest.fn();
    s2.facet.scrollWithAnimation(
      {
        offsetX: {
          value: 10,
          animate: true,
        },
        offsetY: {
          value: 10,
          animate: true,
        },
      },
      10,
      onScrollFinish,
    );
    await sleep(30);

    expect(s2.facet.getScrollOffset()).toStrictEqual({
      scrollY: 10,
      scrollX: 10,
      hRowScrollX: 0,
    });
    expect(onScrollFinish).toBeCalled();
  });
});
