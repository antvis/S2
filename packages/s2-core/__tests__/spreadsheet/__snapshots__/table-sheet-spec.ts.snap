// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder 1`] = `
Object {
  "height": 562,
  "stroke": CSSRGB {
    "alpha": 1,
    "b": 253,
    "colorSpace": "rgb",
    "g": 233,
    "isNone": false,
    "r": 224,
  },
  "strokeOpacity": 1,
  "strokeWidth": 1,
  "width": 799,
  "x": 0,
  "y": 32,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder 2`] = `
Object {
  "fill": "",
  "height": 41,
  "margin": Object {
    "bottom": 24,
    "left": 0,
    "right": 0,
    "top": 0,
  },
  "name": "Empty",
  "width": 64,
  "x": 367.5,
  "y": 292.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder 3`] = `
Object {
  "fill": CSSRGB {
    "alpha": 1,
    "b": 0,
    "colorSpace": "rgb",
    "g": 0,
    "isNone": false,
    "r": 0,
  },
  "fontFamily": "Roboto, PingFangSC, Microsoft YaHei, Arial, sans-serif",
  "fontSize": 12,
  "fontStyle": "normal",
  "fontVariant": "normal",
  "fontWeight": "normal",
  "lineWidth": 1,
  "opacity": 1,
  "text": "暂无数据",
  "x": 375.5,
  "y": 357.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom col height 1`] = `
Object {
  "height": 272,
  "stroke": CSSRGB {
    "alpha": 1,
    "b": 253,
    "colorSpace": "rgb",
    "g": 233,
    "isNone": false,
    "r": 224,
  },
  "strokeOpacity": 1,
  "strokeWidth": 1,
  "width": 599,
  "x": 0,
  "y": 202,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom col height 2`] = `
Object {
  "fill": "",
  "height": 41,
  "margin": Object {
    "bottom": 24,
    "left": 0,
    "right": 0,
    "top": 0,
  },
  "name": "Empty",
  "width": 64,
  "x": 267.5,
  "y": 317.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom col height 3`] = `
Object {
  "fill": CSSRGB {
    "alpha": 1,
    "b": 0,
    "colorSpace": "rgb",
    "g": 0,
    "isNone": false,
    "r": 0,
  },
  "fontFamily": "Roboto, PingFangSC, Microsoft YaHei, Arial, sans-serif",
  "fontSize": 12,
  "fontStyle": "normal",
  "fontVariant": "normal",
  "fontWeight": "normal",
  "lineWidth": 1,
  "opacity": 1,
  "text": "暂无数据",
  "x": 275.5,
  "y": 382.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom col width 1`] = `
Object {
  "height": 442,
  "stroke": CSSRGB {
    "alpha": 1,
    "b": 253,
    "colorSpace": "rgb",
    "g": 233,
    "isNone": false,
    "r": 224,
  },
  "strokeOpacity": 1,
  "strokeWidth": 1,
  "width": 510,
  "x": 0,
  "y": 32,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom col width 2`] = `
Object {
  "fill": "",
  "height": 41,
  "margin": Object {
    "bottom": 24,
    "left": 0,
    "right": 0,
    "top": 0,
  },
  "name": "Empty",
  "width": 64,
  "x": 223,
  "y": 232.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom col width 3`] = `
Object {
  "fill": CSSRGB {
    "alpha": 1,
    "b": 0,
    "colorSpace": "rgb",
    "g": 0,
    "isNone": false,
    "r": 0,
  },
  "fontFamily": "Roboto, PingFangSC, Microsoft YaHei, Arial, sans-serif",
  "fontSize": 12,
  "fontStyle": "normal",
  "fontVariant": "normal",
  "fontWeight": "normal",
  "lineWidth": 1,
  "opacity": 1,
  "text": "暂无数据",
  "x": 231,
  "y": 297.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom icon and description 1`] = `
Object {
  "height": 562,
  "stroke": CSSRGB {
    "alpha": 1,
    "b": 253,
    "colorSpace": "rgb",
    "g": 233,
    "isNone": false,
    "r": 224,
  },
  "strokeOpacity": 1,
  "strokeWidth": 1,
  "width": 799,
  "x": 0,
  "y": 32,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom icon and description 2`] = `
Object {
  "fill": "",
  "height": 41,
  "margin": Object {
    "bottom": 24,
    "left": 0,
    "right": 0,
    "top": 0,
  },
  "name": "Trend",
  "width": 64,
  "x": 367.5,
  "y": 292.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom icon and description 3`] = `
Object {
  "fill": CSSRGB {
    "alpha": 1,
    "b": 0,
    "colorSpace": "rgb",
    "g": 0,
    "isNone": false,
    "r": 0,
  },
  "fontFamily": "Roboto, PingFangSC, Microsoft YaHei, Arial, sans-serif",
  "fontSize": 12,
  "fontStyle": "normal",
  "fontVariant": "normal",
  "fontWeight": "normal",
  "lineWidth": 1,
  "opacity": 1,
  "text": "暂无数据",
  "x": 375.5,
  "y": 357.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom icon and description style 1`] = `
Object {
  "height": 562,
  "stroke": CSSRGB {
    "alpha": 1,
    "b": 253,
    "colorSpace": "rgb",
    "g": 233,
    "isNone": false,
    "r": 224,
  },
  "strokeOpacity": 1,
  "strokeWidth": 1,
  "width": 799,
  "x": 0,
  "y": 32,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom icon and description style 2`] = `
Object {
  "fill": "green",
  "height": 100,
  "margin": Object {
    "bottom": 80,
    "left": 0,
    "right": 0,
    "top": 0,
  },
  "name": "Empty",
  "width": 100,
  "x": 349.5,
  "y": 263,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder by custom icon and description style 3`] = `
Object {
  "fill": CSSRGB {
    "alpha": 1,
    "b": 0,
    "colorSpace": "rgb",
    "g": 0,
    "isNone": false,
    "r": 255,
  },
  "fontFamily": "Roboto, PingFangSC, Microsoft YaHei, Arial, sans-serif",
  "fontSize": 20,
  "fontStyle": "normal",
  "fontVariant": "normal",
  "fontWeight": "normal",
  "lineWidth": 1,
  "opacity": 1,
  "text": "暂无数据",
  "x": 359.5,
  "y": 443,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder for en_US lang 1`] = `
Object {
  "height": 562,
  "stroke": CSSRGB {
    "alpha": 1,
    "b": 253,
    "colorSpace": "rgb",
    "g": 233,
    "isNone": false,
    "r": 224,
  },
  "strokeOpacity": 1,
  "strokeWidth": 1,
  "width": 799,
  "x": 0,
  "y": 32,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder for en_US lang 2`] = `
Object {
  "fill": "",
  "height": 41,
  "margin": Object {
    "bottom": 24,
    "left": 0,
    "right": 0,
    "top": 0,
  },
  "name": "Empty",
  "width": 64,
  "x": 367.5,
  "y": 292.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder for en_US lang 3`] = `
Object {
  "fill": CSSRGB {
    "alpha": 1,
    "b": 0,
    "colorSpace": "rgb",
    "g": 0,
    "isNone": false,
    "r": 0,
  },
  "fontFamily": "Roboto, PingFangSC, Microsoft YaHei, Arial, sans-serif",
  "fontSize": 12,
  "fontStyle": "normal",
  "fontVariant": "normal",
  "fontWeight": "normal",
  "lineWidth": 1,
  "opacity": 1,
  "text": "No Data",
  "x": 377.4892578125,
  "y": 357.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder if contains horizontal scrollbar 1`] = `
Object {
  "height": 562,
  "stroke": CSSRGB {
    "alpha": 1,
    "b": 253,
    "colorSpace": "rgb",
    "g": 233,
    "isNone": false,
    "r": 224,
  },
  "strokeOpacity": 1,
  "strokeWidth": 1,
  "width": 799,
  "x": 0,
  "y": 32,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder if contains horizontal scrollbar 2`] = `
Object {
  "fill": "",
  "height": 41,
  "margin": Object {
    "bottom": 24,
    "left": 0,
    "right": 0,
    "top": 0,
  },
  "name": "Empty",
  "width": 64,
  "x": 367.5,
  "y": 292.5,
}
`;

exports[`TableSheet normal spec Empty Placeholder should render empty placeholder if contains horizontal scrollbar 3`] = `
Object {
  "fill": CSSRGB {
    "alpha": 1,
    "b": 0,
    "colorSpace": "rgb",
    "g": 0,
    "isNone": false,
    "r": 0,
  },
  "fontFamily": "Roboto, PingFangSC, Microsoft YaHei, Arial, sans-serif",
  "fontSize": 12,
  "fontStyle": "normal",
  "fontVariant": "normal",
  "fontWeight": "normal",
  "lineWidth": 1,
  "opacity": 1,
  "text": "暂无数据",
  "x": 375.5,
  "y": 357.5,
}
`;
